OSD-19729: Enables ocm backplane login to provide  default namespace as an argument

amej opened the pull request:
### What type of PR is this?
- feature
- documentation
- unit-test
### What this PR does / Why we need it?
As a SRE, I want backplane login to prompt for a  default namespace for me to run commands like oc debug node/xxx.
### Which Jira/Github issue(s) does this PR fix?

_Resolves #_[OSD-19729](https://issues.redhat.com//browse/OSD-19729)

### Special notes for your reviewer
Once merged,  `./ocm-backplane login --help`  will display                                                                                                                                                                          
```
Running login command will send a request to backplane api
...
Usage:
  ocm-backplane login <CLUSTERID|EXTERNAL_ID|CLUSTER_NAME|CLUSTER_NAME_SEARCH> [flags]
...
Flags:
     -n,  --namespace string   Sets a default namespace for SRE to run commands like oc debug node/xxx
...  
Example: 
Considering cluster name as asathe-19729     

- Without options

```
./ocm-backplane login asathe-19729
oc project
```
Output will be

> _Using project "default" on server "https://api.stage.backplane.openshift.com/backplane/cluster/2afh53rispkj3vbsp6asa4f7hlpb78uv/"._

- By passing a valid kubernetes namespace to the option ' **_namespace_** '
```
./ocm-backplane login --namespace openshift-sre-pruning   28lje9u3jd59j3dmf36sojgpa32qrag3
oc project
```
Output will be
`_Using project "openshift-backplane-srep" on server "https://api.stage.backplane.openshift.com/backplane/cluster/2afh53rispkj3vbsp6asa4f7hlpb78uv/"._`

- By passing an **invalid  namespace** to the option ' **_namespace_** '
```
./ocm-backplane login --namespace ""   asathe-19729
```

Following error  displayed.
_ERRO[0016]  is not a valid namespace_

```
./ocm-backplane login --namespace 'kube_system'   asathe-19729
```
Following error  displayed.                                                                                                                                
ERRO[0012] kube_system is not a valid namespace
```
Hence, the command "**oc status**" will display the following error
_error: Missing or incomplete configuration info.  Please point to an existing, complete config file:_


Please do refer to  the prior discussion in the https://github.com/openshift/backplane-cli/pull/315  .

### Pre-checks (if applicable)

- [x] Ran unit tests locally
- [x] Validated the changes in a cluster
- [x] Included documentation changes with PR

amej commented:
@Tafhim , @supreeth7  and @devppratik : Please can you review ?

samanthajayasinghe commented on cmd/ocm-backplane/login/login.go:
```diff
@@ -324,7 +333,7 @@ func runLogin(cmd *cobra.Command, argv []string) (err error) {
 	targetContext.AuthInfo = targetUserNickName
 	targetContext.Cluster = clusterName
-	targetContext.Namespace = "default"
+	targetContext.Namespace = args.defaultNamespace
```
Small suggestion,  `args.defaultNamespace` may need to validate for empty NS

amej replied:
Thank you.  I can make validation more robust  by ensuring the following [statement](https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/#namespaces-and-dns) is satisfied.
_As a result, all namespace names must be valid [RFC 1123 DNS labels](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-label-names)._

amej replied:
https://github.com/openshift/backplane-cli/pull/374/commits/0560e9731307617ad6feca43106fe01d2d55b398 accomdates your comment.

samanthajayasinghe commented:
@amej Can you pls squash the commits ?

devppratik commented:
/label tide/merge-method-squash

amej commented:
> /label tide/merge-method-squash

@devppratik :  Does this label free me up from the work  of squashing commits locally on my laptop  & force pushing them to the fork's feature branch?

amej reviewed (commented):
I have reviewed the comments. https://github.com/openshift/backplane-cli/pull/374/commits/0560e9731307617ad6feca43106fe01d2d55b398 contains the recommendations.

supreeth7 commented on cmd/ocm-backplane/login/login.go:
```diff
@@ -494,3 +512,11 @@ func getBackplaneCNAME(backplaneURL string) (string, error) {
+func isValidKubernetesNamespace(namespace string) bool {
+	// RFC 1123 compliant regex pattern)
+	pattern := `^[a-z0-9]([-a-z0-9]*[a-z0-9])?$`
+	matched, _ := regexp.MatchString(pattern, namespace)
+	return matched
```
```suggestion
	return regexp.MustCompile(pattern).MatchString(namespace)
```
nit: fewer variable assignments >>

supreeth7 commented on cmd/ocm-backplane/login/login.go:
```diff
@@ -324,7 +334,15 @@ func runLogin(cmd *cobra.Command, argv []string) (err error) {
 	targetContext.Cluster = clusterName
-	targetContext.Namespace = "default"
+
+	logger.Debugln("Validating argument passed as namespace")
```
Shouldn't this only be logged if `args.defaultNamespace!="default"`?

amej replied:
> Shouldn't this only be logged if `args.defaultNamespace!="default"`?

Please review https://github.com/openshift/backplane-cli/pull/374/commits/bce3a48d8838f38d50a2539a948b9f13e9d70cdd .

supreeth7 commented on cmd/ocm-backplane/login/login.go:
```diff
@@ -92,6 +94,14 @@ func init() {
+		"namespace",
+		"n",
+		"default",
+		"Prompts for a  default namespace for SRE to run commands like oc debug node/xxx",
```
Let's be more generic here, Backplane isn't used by only SREs :)

```suggestion
		"The  default namespace for a user to execute commands in",
```

amej commented:
> squash
@supreeth7 : 
As my PR was showing 34 commits, on my current branch OSD-19729; I performed ' git rebase HEAD~34.
It showed me 54 items.  I 'pick' first row & rest  was 'squash'  .  I had to fix multiple merge commits during the rebase.
After rebasing my local branch OSD-19729  on top of the github based OSD-19729, I could push the changes.
Here is the current status with 35 commits : https://github.com/openshift/backplane-cli/pull/374/commits .

amej reviewed (commented):
@supreeth7 : Changes reviewed and action taken including squashing commits locally & pushing them.

supreeth7 commented:
@amej thanks for addressing the review comments.

The commit number is too wide for it to be merged, even with the tide plugin squashing it, can you try squashing your commits without any trailing commit history?

amej commented:
> @amej Can you pls squash the commits ?

@supreeth7 :  Please find a single commit now.  I force pushed the changes to the OSD-19729 branch.

amej commented:
> @amej thanks for addressing the review comments.
> 
> The commit number is too wide for it to be merged, even with the tide plugin squashing it, can you try squashing your commits without any trailing commit history?

@devppratik  guided me to opt for the easy solution .
I opted to force push my changes to the OSD-19729 branch.
Now you will find a single commit .
Thank you for your patience.

supreeth7 reviewed (approved):
/lgtm
